### PR TITLE
Made convolve work with units, closing issue #10811

### LINKED ISSUE #10811: Make convolve work with units
### Description

It would be great if one could properly convolve arrays with units. Currently, the units are stripped off:
```python
>>> import numpy as np
>>> import astropy.units as u
>>> from astropy.convolution import convolve
>>> convolve(np.array([1, 4, 5, 6, 5, 7, 8])*u.ph, [0.2, 0.6, 0.2])
array([1.4, 3.6, 5. , 5.6, 5.6, 6.8, 6.2])
```
As the workaround here is simple (`convolve(a.value, kernel) * a.unit`), it is not urgent at all. User code would ofcourse look better if this would be handled by convolve itself.

### PR DESCRIPTION
**Pull request by george-galvin**
### Description

Both convolve and convolve_fft now output a Quantity array, if the user
 input is one.

EDIT: Fixes #10811

### CODE REVIEW

File: astropy/convolution/convolve.py
Comment (pllim) [inline astropy/convolution/convolve.py:193]:
```suggestion
    # Get array quantity if exists
```

File: astropy/convolution/convolve.py
Comment (pllim) [inline astropy/convolution/convolve.py:377]: Does this make a copy, and thus contradict the comment above it?

**Review by pllim — COMMENTED** [review]
Thanks! Would need test(s).

I am also not very sure about the unnecessary multiplication when `passed_quantity` is 1. Perhaps some benchmark on performance comparing this patch and `master`? I'll defer to `convolution` maintainer(s)'s decisions.

File: astropy/convolution/convolve.py
Comment (george-galvin) [reply astropy/convolution/convolve.py:377]: Thanks for the review! I agree the multiplication by 1 looks redundant - the only alternative I can see is adding a conditional to each return statement, unless there's another way? As for the *=, I wasn't aware that created another copy?

Comment (pllim) [reply astropy/convolution/convolve.py:377]: I am not sure either. I think that depends on the implementation details in `__rmul__`.

Comment (keflavich) [reply astropy/convolution/convolve.py:377]: `*=` should be inplace, so I'm not worried about this, but line 373 above might create a copy.

Returning a ternary (`new_result * passed_quantity if <passed_quantity is a unit> else new_result`) might be better than this unneeded multiplication operation

File: astropy/convolution/convolve.py
Comment (keflavich) [inline astropy/convolution/convolve.py:373]: this likely does create a copy

**Review by mhvk — COMMENTED** [review]
The present method is not bad, but it does needlessly multiply an array with `1` if a non-quantity is passed in.

My sense would be something like:
```
unit = getattr(passed_array, "unit", None)
...
if unit is not None:
    result = result << unit
```
The latter just creates a view and attaches the unit.

Alternatively, since we support `__array_function__`, one could follow what is already done for `np.convolve` (https://github.com/numpy/numpy/blob/2262a7d7a961b1e9322ac1088a18283ed8ec32a0/numpy/core/numeric.py#L754), i.e., introduce a wrapper and define the quantity handler in `units/quantity_helpers/function_helpers.py`, following https://github.com/astropy/astropy/blob/625c6e6257af40e4b5c1070c9b3f707e7b0953dd/astropy/units/quantity_helper/function_helpers.py#L578-L583

The latter has the advantage that if, e.g., someone downstream wants to use `convolve` with their own array-like class (say, dask), then they can easily override the behaviour.

Looking at the code for `np.convolve` also reminds me: in principle, the kernel can have a unit too...

**Comment by george-galvin** [comment]
Okay, I've implemented @mhvk's solution, which is way nicer looking and solves all the problems above.

**Review by pllim — COMMENTED** [review]
Still need test(s). And perhaps squash the commits? Thanks!

**Comment by george-galvin** [comment]
Ok, added two tests and squashed into one commit

File: astropy/convolution/tests/test_convolve_fft.py
Comment (pllim) [inline astropy/convolution/tests/test_convolve_fft.py:73]: Does this need a `@pytest.mark.parametrize(option_names, options)` like the other tests?

File: astropy/convolution/tests/test_convolve.py
Comment (pllim) [inline astropy/convolution/tests/test_convolve.py:68]: Does this need parameterization like the other tests?
```
    @pytest.mark.parametrize(('boundary', 'nan_treatment',
                              'normalize_kernel', 'preserve_nan', 'dtype'),
                             itertools.product(BOUNDARY_OPTIONS,
                                               NANHANDLING_OPTIONS,
                                               NORMALIZE_OPTIONS,
                                               PRESERVE_NAN_OPTIONS,
                                               VALID_DTYPES))
```

Comment (george-galvin) [reply astropy/convolution/tests/test_convolve.py:68]: I used the first couple of functions on test_convolve.py as a template, and so thought parameterizing it would be unnecessary - but this could be added instead of the equality check.

Comment (pllim) [reply astropy/convolution/tests/test_convolve.py:68]: @keflavich , what do you think?

Comment (keflavich) [reply astropy/convolution/tests/test_convolve.py:68]: I'd default to parametrizing the same way.  In principle, there's no interaction between what's being tested and the other parameters, but that's sort of the point of the test - make sure that's true and we never regress.

**Comment by pllim** [comment]
Thanks for your patience! RTD and CircleCI failures don't seem related. I restarted Travis CI, just in case.

@keflavich and @mhvk , would you be able to do a final review and approve if it looks fine?

File: astropy/convolution/convolve.py
Comment (mhvk) [inline astropy/convolution/convolve.py:194]: Nitpicky, but I think it would be clearer in this variable is called `array_unit` (or something similar)

File: astropy/convolution/convolve.py
Comment (mhvk) [inline astropy/convolution/convolve.py:359]: Looking at this again, I think since we keep on purpose a reference to the input array, we might as well look for the unit only here, i.e., something like
```
unit = getattr(passed_array, 'unit', None)
if unit is not None:
    result <<= unit
```
(and the same for the other routine)

**Review by mhvk — COMMENTED** [review]
Looks good modulo a suggestion to put things together (which I think is possible; if that is wrong, then a suggestion for a name change to make things a bit clearer)

**Comment by mhvk** [comment]
p.s. The travis windows test failure also seems unrelated - what is happening!?!?

**Comment by pllim** [comment]
Re: Windows CI -- No idea! Started happening today or so. Trying things out at #10842

File: astropy/convolution/convolve.py
Comment (george-galvin) [reply astropy/convolution/convolve.py]: Good points! In convolve_fft a reference to the input array isn't kept, so I didn't change the positioning of the line which looks for the unit. I've changed everything else you pointed out though.

**Review by mhvk — APPROVED** [review]
Thanks! Looks good to me.

**Comment by george-galvin** [comment]
Thanks! Does this get merged now?

**Comment by mhvk** [comment]
Yes, it will get merged! I just thought I'd give others a chance to have a last look, and then forgot... Thanks!!

**Comment by olebole** [comment]
Thank you very much @george-galvin! I am very impressed by how quick my feature request is resolved! That is great work!

**Comment by george-galvin** [comment]
Great, thanks for helping with my first PR guys :)

**Pull request merged**
